probability/windowed_median.cpp

```diff
@@ -0,0 +1,226 @@
+     * between the two values in the middle
+     */
+    float getMedian() const {
+        if (_sortedValues.size() % 2 != 0) {
+            return *_itMedian;  // O(1)
+        }
+        return 0.5f * *_itMedian + 0.5f * *next(_itMedian);  // O(1)
+    }
+
+    /**
+     * @brief A naive and inefficient method to obtain the median of the sliding
+     * window. Used for testing!
+     * @return Median of sliding window. For even window size return the average
+     * between the two values in the middle
+     */
+    float getMedianNaive() const {
+        auto window = _window;
+        window.sort();  // Sort window - O(NlogN)
+        auto median =
+            *next(window.begin(),
+                  window.size() / 2);  // Find value in the middle - O(N)
+        if (window.size() % 2 != 0) {
+            return median;
+        }
+        return 0.5f * median +
+               0.5f * *next(window.begin(), window.size() / 2 - 1);  // O(N)
+    }
+};
+}  // namespace windowed_median
+}  // namespace probability
+
+/**
+ * @brief Self-test implementations
+ * @param vals Stream of values
+ * @param windowSize Size of sliding window
+ */
+static void test(const std::vector<int> &vals, int windowSize) {
+    probability::windowed_median::WindowedMedian windowedMedian(windowSize);
+    for (const auto val : vals) {
+        windowedMedian.insert(val);
+
+        // Comparing medians: efficient function vs. Naive one
+        assert(windowedMedian.getMedian() == windowedMedian.getMedianNaive());
+    }
+}
+
+/**
+ * @brief Main function
+ * @param argc command line argument count (ignored)
+ * @param argv command line array of arguments (ignored)
+ * @returns 0 on exit
+ */
+int main(int argc, const char *argv[]) {
+    
+    // A few fixed test cases
+    test({1, 2, 3, 4, 5, 6, 7, 8, 9}, 3);   // Array of sorted values; odd window size
+    test({9, 8, 7, 6, 5, 4, 3, 2, 1}, 3);   // Array of sorted values - decreasing; odd window size
+    test({9, 8, 7, 6, 5, 4, 5, 6}, 4);      // Even window size
+    test({3, 3, 3, 3, 3, 3, 3, 3, 3}, 3);   // Array with repeating values
+    test({3, 3, 3, 3, 7, 3, 3, 3, 3}, 3);   // Array with same values except one
+    test({4, 3, 3, -5, -5, 1, 3, 4, 5}, 5); // Array that includes repeating values including negatives
+    
+    // Array with large values - sum of few pairs exceeds MAX_INT. Window size is even - testing calculation of
+    // average median between two middle values
+    test({470211272, 101027544, 1457850878, 1458777923, 2007237709, 823564440,
+          1115438165, 1784484492, 74243042, 114807987}, 6);
+    
+    // Random test cases
+    std::srand(static_cast<unsigned int>(std::time(nullptr)));
+    std::vector<int> vals;
+    for (int i = 8; i < 100; i++) {
+        const auto n = 1 + std::rand() / ((RAND_MAX + 5u) / 20);    // Array size in the range [5, 20]
+        auto windowSize = 1 + std::rand() / ((RAND_MAX + 3u) / 10); // Window size in the range [3, 10]
```

Comment:
Please use `///` here. Same in other parts of the code. 🙂
```suggestion
        const auto n = 1 + std::rand() / ((RAND_MAX + 5u) / 20);    /// Array size in the range [5, 20]
        auto windowSize = 1 + std::rand() / ((RAND_MAX + 3u) / 10); /// Window size in the range [3, 10]
```

Reply:
OK, thank you. It's a learning curve for me; I'll have to adjust :)